src/ucs/sys/event_set.c

```diff
@@ -134,36 +135,57 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
     return UCS_OK;
 }
 
-ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
-                                ucs_event_set_handler_t event_set_handler,
-                                void *arg)
+unsigned ucs_event_set_wait(ucs_sys_event_set_t *event_set, int max_events,
+                            int timeout_ms,
+                            ucs_event_set_handler_t event_set_handler,
+                            void *arg)
 {
-    int nready;
+    int read_events = 0;
     struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
-    int i;
-
-    if (event_set_handler == NULL) {
-        return UCS_ERR_INVALID_PARAM;
-    }
-
-    nready = epoll_wait(event_set->epfd, ep_events, UCS_EVENT_EPOLL_MAX_EVENTS,
-                        timeout_ms);
-    if ((nready < 0) && (errno != EINTR)) {
-        ucs_error("epoll_wait() failed: %m");
-        return UCS_ERR_IO_ERROR;
-    }
-    ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
-                   event_set->epfd, timeout_ms, nready);
-
-    for (i=0; i < nready; i++) {
-        int events = ucs_event_set_map_to_events(ep_events[i].events);
-        event_set_handler(ep_events[i].data.fd, events, arg);
-    }
-    return UCS_OK;
+    int i, max_wait_events, nready;
+
+    ucs_assert(event_set_handler != NULL);
+
+    do {
+        max_wait_events = ucs_min(max_events - read_events,
+                                   UCS_EVENT_EPOLL_MAX_EVENTS);
+	nready = epoll_wait(event_set->epfd, ep_events,
+                            max_wait_events, timeout_ms);
+	if (nready < 0) {
+            if (errno == EINTR) {
+                /* here we should take care about timeout_ms (i.e. decrease it),
+                 * but it is ok to not od this */
+                continue;
+            }
+//            ucs_error("epoll_wait() failed: %m");
```

Comment:
How to care errors except `EINTR`. What do you think to return negative value(-1) in that case?

Reply:
it should return `0` as a function return value, and we can do something like:
```
unsigned ucs_event_set_wait(ucs_sys_event_set_t *event_set, int max_events,
                            int timeout_ms,
                            ucs_event_set_handler_t event_set_handler,
                            void *arg, ucs_status_t *status_p)
```
...
```
ucs_error(...);
if (status_p != NULL) {
    *status_p = UCS_ERR_IO_ERROR;
}
```

if a user needs to handle errors, it passes storage for `status`, otherwise - `NULL`

Reply:
What do you think to return `UCS_OK` on success instead of NULL.
(Or reset NULL on the beginning function.)

I use `status` value on multiple places.
I'm worrying to use the previus result.

```C
    status = ucs_event_set_add(event_set, pipefd[0], UCS_EVENT_SET_EVREAD,
                               NULL);
    EXPECT_EQ(UCS_OK,status);

    nread = ucs_event_set_wait(event_set, MAX_EVENT_SET_SIZE, 20,
                               event_set_func3, NULL, &status);
```

Reply:
I'm so sorry for being unclear
I meant that we should set `UCS_OK` on success and `UCS_ERR_IO_ERROR` in case of any error occurred if user specified `status_p`, otherwise, we should do nothing